src/drafty.js

```diff
@@ -597,28 +599,61 @@ Drafty.parse = function(content) {
     result.txt = blx[0].txt;
     result.fmt = (blx[0].fmt || []).concat(blx[0].ent || []);
 
+    if (result.fmt.length) {
+      const segments = segmenter.segment(result.txt);
+      for (const ele of result.fmt) {
+        ({
+            at: ele.at,
+            len: ele.len
+          } =
+          getCorrectLengthsWhenTreatedAsGrapheme(ele, segments, result.txt));
+      }
+    }
+
     for (let i = 1; i < blx.length; i++) {
       const block = blx[i];
-      const offset = result.txt.length + 1;
+      const offset = getGraphemesFromString(result.txt).length + 1;
 
       result.fmt.push({
         tp: 'BR',
         len: 1,
         at: offset - 1
       });
 
+      let segments = {};
+
       result.txt += ' ' + block.txt;
       if (block.fmt) {
-        result.fmt = result.fmt.concat(block.fmt.map((s) => {
-          s.at += offset;
-          return s;
-        }));
+        segments = segmenter.segment(block.txt);
+        result.fmt = result.fmt.concat(
+          block.fmt.map((s) => {
+            const {
```

Comment:
You use `({` above. I suggest to choose one and stick to it: either `const {` or `({`. I like `const {` more.